### PR TITLE
[website] Add TPC-DS benchmark documentation

### PR DESCRIPTION
**Pull request by aromanenko-dev**
Add a dedicated page for TPC-DS benchmark, like we have for Nexmark one.
Issue #21674 

R: @TheNeuralBit 
R: @echauchot 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by aromanenko-dev** [comment]
@TheNeuralBit Could you take a look as well (I asked @echauchot too) as a native English speaker for some grammar errors?

**Comment by TheNeuralBit** [comment]
yeah I can take a look, sorry for the delay

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+- Detect missing Beam SQL features or incompatibilities
+- Find performance issues in Beam
+
+### Queries
+All TPC-DS queries in Beam are pre-generated and stored in the provided artifacts.
+
+For the moment, 28 out of 103 SQL queries (99 + 4) successfully pass by running with Beam SQL transform since not all
+SQL-99 operations are supported.
+
+Currently supported queries:
```
Comment (echauchot) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:69]: add "are"

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+> purpose decision support system."
+
+- Industry standard benchmark (OLAP/Data Warehouse)
+  - http://www.tpc.org/tpcds/
+- Implemented for many analytical processing systems - RDBMS, Apache Spark, Apache Flink, etc
+- Wide range of different queries (SQL)
+- Existing tools to generate input data of different sizes
+
+## Table schemas
+TBD
```
Comment (echauchot) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:34]: extracting the used tables from all the TPCDS table would be painful so maybe just give a word about the different tables like this: 
**store_sales**: contains the sell records of a virtual store

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+### Runtime
+TPC-DS extension for Beam can only be run in **Batch** mode and supports these runners for the moment:
+- Spark Runner
+- Flink Runner
+- Dataflow Runner
+
+## TPC-DS output
+
+TBD
```
Comment (echauchot) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:86]: for now just talk about the response time in the output log

**Review by echauchot — CHANGES_REQUESTED** [review]
Thanks Alexey for this update of the docs. Only minor missing points

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md (hunk 2 above)
Comment (TheNeuralBit) [reply website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:34]: +1

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+## Table schemas
+TBD
+
+## The queries
+
+TPC-DS benchmark contains 99 distinct SQL-99 queries (including OLAP extensions). Each query answers a business
+question, which illustrates the business context in which the query could be used.
+
+All queries are “templated” with random input parameters and used to compare SQL implementation of completeness and
+performance.
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:42]:
```suggestion
All queries are “templated” with random input parameters and used to compare completeness and
performance of SQL implementations. 
```

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+- Input files (CSV) are generated with CLI tool `dsdgen`
+- Input datasets can be generated for different sizes:
+  - 1GB / 10GB / 100GB / 1000GB
+- The tool constraints the minimum amount of data to be generated to 1GB
+
+## TPC-DS extension in Beam
+
+### Reasons
+
+Beam provides a simplified implementation of TPC-DS benchmark and there are several reasons to have it in Beam:
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:56]:
```suggestion
There are several reasons to have TPC-DS benchmarks in Beam:
```

I think the first part of this belongs somewhere else, maybe in a summary under "TPC-DS extension in Beam", which also links to the location in the repo?

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+SQL-99 operations are supported.
+
+Currently supported queries:
+ - 3, 7, 10, 22, 25, 26, 29, 35, 38, 40, 42, 43, 50, 52, 55, 69, 78, 79, 83, 84, 87, 93, 96, 97, 99
+
+### Tables
+All TPC-DS table schemas are stored in the provided artifacts.
+
+### Input data
+Input data is already pre-generated for two data formats (CSV and Parquet) and stored in Google Cloud Storage (gs://beam-tpcds)
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:76]:
```suggestion
CSV and Parquet input data has been pre-generated and staged in the in the Google Cloud Storage bucket `gs://beam-tpcds`.
```

You might add a summary of the directory structure here, like:
```
gs://beam-tpcds/datasets/{parquet,text}/{partitioned,nonpartitioned}/{1GB,10GB,100GB}/<table name>
```
with a brief explanation (for example presumably text = csv, but that's not immediately obvious, it's also unclear what the partitioned directory contains).

A couple of other thoughts:
- Maybe add a README in gs://beam-tpc-ds linking back here
- Do we need all of datasets, gcpTempLocation, output, resources, results in that bucket?

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        `settings.gradle.kts`.
+
+Test data has to be generated before running a suite and stored to accessible file system. The query results will be written into output files.
+
+### Common configuration parameters
+
+Size of input dataset (1GB / 10GB / 100GB / 1000GB):
+
+    --dataSize=<1GB|10GB|100GB|1000GB>
+
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:111]:
```suggestion
- ` --dataSize=<1GB|10GB|100GB|1000GB>`: Size of input dataset
```

you might structure this as a bulleted list

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+Path to results directory:
+
+    --resultsDirectory=<path to dir>
+
+Format of input files:
+
+    --sourceType=<CSV|PARQUET>
+
+Run queries (comma separated list of query numbers or `all` for all queries):
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:124]:
```suggestion
Select queries to run (comma separated list of query numbers or `all` for all queries):
```

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+    --queries=<1,2,...N|all>
+
+Number of queries **N** that are running in parallel:
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:128]:
```suggestion
Number of queries **N** to run in parallel:
```

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+            --autoscalingAlgorithm=NONE
+            --dataSize=100GB
+            --sourceType=PARQUET
+            --dataDirectory=gs://beam-tpcds/datasets/parquet/partitioned
+            --resultsDirectory=/tmp/beam-tpcds/results/dataflow/
+            --tpcParallel=4
+            --queries=all"
+
+## TPC-DS dashboards
+TBD
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:182]: I don't think we should publish this with the "TBD" sections still here, do you intend to address these? If not we could just comment them out for now so they don't show up on the website.

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+    --queries=<1,2,...N|all>
+
+Number of queries **N** that are running in parallel:
+
+    --tpcParallel=N
+
+## Running TPC-DS
+
+There are some examples how to run TPC-DS benchmark on different runners.
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:134]:
```suggestion
Here are some examples demonstrating how to run TPC-DS benchmarks on different runners.
```

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+Running suite on the SparkRunner (local) with Query3 against 1Gb dataset in Parquet format:
+
+    ./gradlew :sdks:java:testing:tpcds:run \
+        -Ptpcds.runner=":runners:spark:3" \
+        -Ptpcds.args="
+            --runner=SparkRunner
+            --dataSize=1GB
+            --sourceType=PARQUET
+            --dataDirectory=gs://beam-tpcds/datasets/parquet/partitioned
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:144]: Could/should we resolve this path given dataSize and sourceType? (we could still allow the user to override it)

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+            --runner=FlinkRunner
+            --parallelism=2
+            --dataSize=10GB
+            --sourceType=CSV
+            --dataDirectory=gs://beam-tpcds/datasets/csv
+            --resultsDirectory=/tmp/beam-tpcds/results/flink/
+            --tpcParallel=2
+            --queries=7,10"
+
+Running suite on the DataflowRunner (local) with all queries against 100Gb dataset in PARQUET format:
```
[the remark this answers is not in the document]
Comment (TheNeuralBit) [reply website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:163]: Also what is meant by "local" here, I'm assuming that's a copy paste error?

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+            --runner=DataflowRunner
+            --region=<region_name>
+            --project=<project_name>
+            --numWorkers=4
+            --maxNumWorkers=4
+            --autoscalingAlgorithm=NONE
+            --dataSize=100GB
+            --sourceType=PARQUET
+            --dataDirectory=gs://beam-tpcds/datasets/parquet/partitioned
+            --resultsDirectory=/tmp/beam-tpcds/results/dataflow/
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:177]: Should the be a cloud storage path?

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+## TPC-DS extension in Beam
+
+### Reasons
+
+Beam provides a simplified implementation of TPC-DS benchmark and there are several reasons to have it in Beam:
+
+- Compare the performance boost or degradation of Beam SQL for different runners or their versions
+- Run Beam SQL on different runtime environments
+- Detect missing Beam SQL features or incompatibilities
+- Find performance issues in Beam
```
Comment (TheNeuralBit) [inline website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:61]:
```suggestion
- Compare the performance of Beam SQL against native SQL implementations for different runners.
- Exercise Beam SQL on different runtime environments.
- Identify missing or incorrect Beam SQL features.
- Identify performance issues in Beam and Beam SQL.
```

**Review by TheNeuralBit — APPROVED** [review]
Thanks @aromanenko-dev!

FYI @kennknowles, @apilloud

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md (hunk 2 above)
Comment (aromanenko-dev) [reply website/www/site/content/en/documentation/sdks/java/testing/tpcds.md]: Agree, done

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md (hunk 3 above)
Comment (aromanenko-dev) [reply website/www/site/content/en/documentation/sdks/java/testing/tpcds.md]: I added an example of output

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+
+### Queries
+All TPC-DS queries in Beam are pre-generated and stored in the provided artifacts.
+
+For the moment, 28 out of 103 SQL queries (99 + 4) successfully pass by running with Beam SQL transform since not all
+SQL-99 operations are supported.
+
+Currently supported queries:
+ - 3, 7, 10, 22, 25, 26, 29, 35, 38, 40, 42, 43, 50, 52, 55, 69, 78, 79, 83, 84, 87, 93, 96, 97, 99
```
[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply website/www/site/content/en/documentation/sdks/java/testing/tpcds.md:106]: Not sure if it's easy to add, though I guess we may keep a list of such queries somewhere. For now, I just added a comment `as of ...`

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md (hunk 6 above)
Comment (aromanenko-dev) [reply website/www/site/content/en/documentation/sdks/java/testing/tpcds.md]: I added some details about `beam-tpcds` bucket structure. README will be updates separately. All other buckets are needed to run a benchmark.

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md (hunk 7 above)
Comment (aromanenko-dev) [reply website/www/site/content/en/documentation/sdks/java/testing/tpcds.md]: Could you give an example? Not sure that I get it...

File: website/www/site/content/en/documentation/sdks/java/testing/tpcds.md (hunk 10 above)
Comment (aromanenko-dev) [reply website/www/site/content/en/documentation/sdks/java/testing/tpcds.md]: Commented for now. I will add this section once it will be done in Grafana

**Comment by aromanenko-dev** [comment]
@echauchot @TheNeuralBit  Thanks for review! I tried to address all your comments, ptal

**Comment by aromanenko-dev** [comment]
Run Website PreCommit

**Comment by aromanenko-dev** [comment]
Run Website_Stage_GCS PreCommit

**Comment by aromanenko-dev** [comment]
I addressed the comments and I'll self-merge this PR since it was already LGTM-ed.

**Pull request merged**